### ISSUE TITLE
[JENKINS-56341] GIT_COMMIT can point to a no repo commit SHA

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/workflow-multibranch-plugin#502]
<p>We use multibranch pipelines for our project, it works pretty well and it is nice to have PRs tested merged with the master branch, we also use Blueocean for our UI, and here we have to use a workaround to make our UI user-friendly.<br/>
Our jobs spin 100-150 test in parallel, BO is not really user-friendly with that amount of parallel stages so we make groups and launch a downstream job for each group, at this point all OK, the problem starts when you try to pass the GIT_COMMIT<br/>
to the downstream job if the master branch has changed when the checkout make the merge, the resulting commit (GIT_COMMIT env variable) is not in the repo,so the downstream job cannot checkout the GIT_COMMIT and it fails. <br/>
I can pass the GIT_BRANCH but I'd prefer to pass the commit sha1. I start to play with `git rev-list HEAD -4` and `git reflog -6` to see if always have the same pattern and try to grab the correct commit sha1, but I start thinking that should be a better way. <br/>
Also, I can disable the merge and make it on the downstream job but I don't like it either. </p>

<p>GIT_PREVIOUS_COMMIT also point to an existing commit so I cannot use it.</p>

<p>In this example, the PR head commit is `08da92d31512c4a0c24948bc80f911f83c075070`, but GIT_COMMIT points to'7c36975a772dbab59e20169a0ac01e60fb6db739' and GIT_PREVIOUS_COMMIT to '6cd1da59add5d188cad41b9cbdcd33f4a7241c63' none of those in the repo</p>

<pre>
using credential token
 > git rev-parse --is-inside-work-tree # timeout=10
Fetching changes from the remote Git repository
 > git config remote.origin.url https:<span class="code-comment">//github.com/org/repo.git # timeout=10
</span>Fetching without tags
Fetching upstream changes from https:<span class="code-comment">//github.com/org/repo.git
</span> > git --version # timeout=10
using GIT_ASKPASS to set credentials GitHub user @​elasticmachine User + Personal Access Token
 > git fetch --no-tags --progress https:<span class="code-comment">//github.com/org/arepo.git +refs/pull/352/head:refs/remotes/origin/PR-352 +refs/heads/master:refs/remotes/origin/master
</span>Merging remotes/origin/master commit e067dce0923be39f9b48e721240cf26ee083ab9e into PR head commit 08da92d31512c4a0c24948bc80f911f83c075070
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 08da92d31512c4a0c24948bc80f911f83c075070
 > git merge e067dce0923be39f9b48e721240cf26ee083ab9e # timeout=10
 > git rev-parse HEAD^{commit} # timeout=10
Merge succeeded, producing a620f8fbd54f038733b1b238d90c59ccf18f3fc1
Checking out Revision a620f8fbd54f038733b1b238d90c59ccf18f3fc1 (PR-352)
Commit message: <span class="code-quote">"Merge commit <span class="code-quote">'e067dce0923be39f9b48e721240cf26ee083ab9e'</span> into HEAD"</span>
First time build. Skipping changelog.
> git checkout -f 08da92d31512c4a0c24948bc80f911f83c075070
> git merge e067dce0923be39f9b48e721240cf26ee083ab9e # timeout=10
> git rev-parse HEAD^{commit} # timeout=10
> git config core.sparsecheckout # timeout=10
> git checkout -f a620f8fbd54f038733b1b238d90c59ccf18f3fc1
> git rev-list --no-walk 6cd1da59add5d188cad41b9cbdcd33f4a7241c63 # timeout=10
</pre>

<pre>
[2019-02-28T19:36:16.610Z] export GIT_BRANCH=<span class="code-quote">'PR-352'</span>
[2019-02-28T19:36:16.838Z] export GIT_COMMIT=<span class="code-quote">'7c36975a772dbab59e20169a0ac01e60fb6db739'</span>
[2019-02-28T19:36:16.838Z] export GIT_PREVIOUS_COMMIT=<span class="code-quote">'6cd1da59add5d188cad41b9cbdcd33f4a7241c63'</span>
</pre>

<p>I can grab `git HEAD -1` but I am not sure always would be correct because if there aren't changes on the master branch GIT_COMMIT is correct.</p>

<pre>
[2019-02-28T19:36:16.838Z] + git reflog -6
[2019-02-28T19:36:16.838Z] a620f8f HEAD@{0}: merge e067dce0923be39f9b48e721240cf26ee083ab9e: Merge made by the <span class="code-quote">'recursive'</span> strategy.
[2019-02-28T19:36:16.838Z] 08da92d HEAD@{1}: checkout: moving from master to 08da92d31512c4a0c24948bc80f911f83c075070
</pre>

<pre>
[2019-02-28T19:36:16.838Z] + git rev-list HEAD -4
[2019-02-28T19:36:16.838Z] a620f8fbd54f038733b1b238d90c59ccf18f3fc1
[2019-02-28T19:36:16.838Z] 08da92d31512c4a0c24948bc80f911f83c075070
[2019-02-28T19:36:16.838Z] 6646f0310ea4e6e176faa178a6a2ecf3be11168b
[2019-02-28T19:36:16.838Z] 6ff5224c0892801c2632be855d91a38426677ba5
</pre>

<p>the workaround is to use git to calculate the commit </p>

<pre>
def getBaseCommit(){
  def baseCommit = ''
  def latestCommit = sh(label: <span class="code-quote">'Get previous commit'</span>, script: <span class="code-quote">"git rev-parse HEAD"</span>, returnStdout: <span class="code-keyword">true</span>)?.trim()
  def previousCommit = sh(label: <span class="code-quote">'Get previous commit'</span>, script: <span class="code-quote">"git rev-parse HEAD^"</span>, returnStdout: <span class="code-keyword">true</span>)?.trim()
  <span class="code-keyword">if</span>(env?.CHANGE_ID == <span class="code-keyword">null</span>){
    baseCommit = env.GIT_COMMIT
  } <span class="code-keyword">else</span> <span class="code-keyword">if</span>(<span class="code-quote">"${env.GIT_COMMIT}"</span>.equals(<span class="code-quote">"${latestCommit}"</span>)){
    baseCommit = env.GIT_COMMIT
  } <span class="code-keyword">else</span> {
    baseCommit = previousCommit
  }
  env.GIT_BASE_COMMIT = baseCommit
  <span class="code-keyword">return</span> baseCommit
}
</pre>

<p>Google groups conversation about it<br/>
<a href="https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/jenkinsci-dev/Dus0VX6y04g/XwVPA8odBwAJ" class="external-link" target="_blank" rel="nofollow noopener">https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/jenkinsci-dev/Dus0VX6y04g/XwVPA8odBwAJ</a></p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ifernandezcalvo.png" title="ifernandezcalvo's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ifernandezcalvo">ifernandezcalvo</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-56341" target="_blank">GIT_COMMIT can point to a no repo commit SHA</a></i></summary>
<i><ul>
<li><b>status</b>: Open
<li><b>priority</b>: Minor
<li><b>component(s)</b>: workflow-multibranch-plugin
<li><b>resolution</b>: Unresolved
<li><b>votes</b>: 0
<li><b>watchers</b>: 4
<li><b>imported</b>: 20251212-090250
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>We use multibranch pipelines for our project, it works pretty well and it is nice to have PRs tested merged with the master branch, we also use Blueocean for our UI, and here we have to use a workaround to make our UI user-friendly.
Our jobs spin 100-150 test in parallel, BO is not really user-friendly with that amount of parallel stages so we make groups and launch a downstream job for each group, at this point all OK, the problem starts when you try to pass the GIT_COMMIT
to the downstream job if the master branch has changed when the checkout make the merge, the resulting commit (GIT_COMMIT env variable) is not in the repo,so the downstream job cannot checkout the GIT_COMMIT and it fails. 
I can pass the GIT_BRANCH but I'd prefer to pass the commit sha1. I start to play with `git rev-list HEAD -4` and `git reflog -6` to see if always have the same pattern and try to grab the correct commit sha1, but I start thinking that should be a better way. 
Also, I can disable the merge and make it on the downstream job but I don't like it either. </p>

<p>GIT_PREVIOUS_COMMIT also point to an existing commit so I cannot use it.</p>

<p>In this example, the PR head commit is `08da92d31512c4a0c24948bc80f911f83c075070`, but GIT_COMMIT points to'7c36975a772dbab59e20169a0ac01e60fb6db739' and GIT_PREVIOUS_COMMIT to '6cd1da59add5d188cad41b9cbdcd33f4a7241c63' none of those in the repo</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">using credential token
 &gt; git rev-parse --is-inside-work-tree # timeout=10
Fetching changes from the remote Git repository
 &gt; git config remote.origin.url https:<span class="code-comment">//github.com/org/repo.git # timeout=10
</span>Fetching without tags
Fetching upstream changes from https:<span class="code-comment">//github.com/org/repo.git
</span> &gt; git --version # timeout=10
using GIT_ASKPASS to set credentials GitHub user @elasticmachine User + Personal Access Token
 &gt; git fetch --no-tags --progress https:<span class="code-comment">//github.com/org/arepo.git +refs/pull/352/head:refs/remotes/origin/PR-352 +refs/heads/master:refs/remotes/origin/master
</span>Merging remotes/origin/master commit e067dce0923be39f9b48e721240cf26ee083ab9e into PR head commit 08da92d31512c4a0c24948bc80f911f83c075070
 &gt; git config core.sparsecheckout # timeout=10
 &gt; git checkout -f 08da92d31512c4a0c24948bc80f911f83c075070
 &gt; git merge e067dce0923be39f9b48e721240cf26ee083ab9e # timeout=10
 &gt; git rev-parse HEAD^{commit} # timeout=10
Merge succeeded, producing a620f8fbd54f038733b1b238d90c59ccf18f3fc1
Checking out Revision a620f8fbd54f038733b1b238d90c59ccf18f3fc1 (PR-352)
Commit message: <span class="code-quote">"Merge commit <span class="code-quote">'e067dce0923be39f9b48e721240cf26ee083ab9e'</span> into HEAD"</span>
First time build. Skipping changelog.
&gt; git checkout -f 08da92d31512c4a0c24948bc80f911f83c075070
&gt; git merge e067dce0923be39f9b48e721240cf26ee083ab9e # timeout=10
&gt; git rev-parse HEAD^{commit} # timeout=10
&gt; git config core.sparsecheckout # timeout=10
&gt; git checkout -f a620f8fbd54f038733b1b238d90c59ccf18f3fc1
&gt; git rev-list --no-walk 6cd1da59add5d188cad41b9cbdcd33f4a7241c63 # timeout=10
</pre>
</div></div>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">[2019-02-28T19:36:16.610Z] export GIT_BRANCH=<span class="code-quote">'PR-352'</span>
[2019-02-28T19:36:16.838Z] export GIT_COMMIT=<span class="code-quote">'7c36975a772dbab59e20169a0ac01e60fb6db739'</span>
[2019-02-28T19:36:16.838Z] export GIT_PREVIOUS_COMMIT=<span class="code-quote">'6cd1da59add5d188cad41b9cbdcd33f4a7241c63'</span>
</pre>
</div></div>

<p>I can grab `git HEAD -1` but I am not sure always would be correct because if there aren't changes on the master branch GIT_COMMIT is correct.</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">[2019-02-28T19:36:16.838Z] + git reflog -6
[2019-02-28T19:36:16.838Z] a620f8f HEAD@{0}: merge e067dce0923be39f9b48e721240cf26ee083ab9e: Merge made by the <span class="code-quote">'recursive'</span> strategy.
[2019-02-28T19:36:16.838Z] 08da92d HEAD@{1}: checkout: moving from master to 08da92d31512c4a0c24948bc80f911f83c075070
</pre>
</div></div>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">[2019-02-28T19:36:16.838Z] + git rev-list HEAD -4
[2019-02-28T19:36:16.838Z] a620f8fbd54f038733b1b238d90c59ccf18f3fc1
[2019-02-28T19:36:16.838Z] 08da92d31512c4a0c24948bc80f911f83c075070
[2019-02-28T19:36:16.838Z] 6646f0310ea4e6e176faa178a6a2ecf3be11168b
[2019-02-28T19:36:16.838Z] 6ff5224c0892801c2632be855d91a38426677ba5
</pre>
</div></div>

<p>the workaround is to use git to calculate the commit </p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">def getBaseCommit(){
  def baseCommit = ''
  def latestCommit = sh(label: <span class="code-quote">'Get previous commit'</span>, script: <span class="code-quote">"git rev-parse HEAD"</span>, returnStdout: <span class="code-keyword">true</span>)?.trim()
  def previousCommit = sh(label: <span class="code-quote">'Get previous commit'</span>, script: <span class="code-quote">"git rev-parse HEAD^"</span>, returnStdout: <span class="code-keyword">true</span>)?.trim()
  <span class="code-keyword">if</span>(env?.CHANGE_ID == <span class="code-keyword">null</span>){
    baseCommit = env.GIT_COMMIT
  } <span class="code-keyword">else</span> <span class="code-keyword">if</span>(<span class="code-quote">"${env.GIT_COMMIT}"</span>.equals(<span class="code-quote">"${latestCommit}"</span>)){
    baseCommit = env.GIT_COMMIT
  } <span class="code-keyword">else</span> {
    baseCommit = previousCommit
  }
  env.GIT_BASE_COMMIT = baseCommit
  <span class="code-keyword">return</span> baseCommit
}
</pre>
</div></div>

<p>Google groups conversation about it
<a href="https://groups.google.com/forum/?utm_medium=email&amp;utm_source=footer#!msg/jenkinsci-dev/Dus0VX6y04g/XwVPA8odBwAJ" class="external-link" target="_blank" rel="nofollow noopener">https://groups.google.com/forum/?utm_medium=email&amp;utm_source=footer#!msg/jenkinsci-dev/Dus0VX6y04g/XwVPA8odBwAJ</a></p></pre>
</details>
</details>
